HW0/P4/P4.py

```diff
@@ -40,38 +48,132 @@
+        #Load the measurements text file into a matrix
+
+    stretch_matrix = np.matrix([[1/rx, 0, 0], [0, 1/ry, 0], [0, 0, 1/rz]])
+        # Matrix with the stretching factors
+
+    corrected = stretch_matrix * s_meas
+        # Correct position by multiplying the stretch matrix by the measured position
+ 
+    x_coords = np.array(corrected)[0, :]  # extract x coordinates
+    y_coords = np.array(corrected)[1, :]  # extract y coordinates
+    z_coords = np.array(corrected)[2, :]  # extract z coordinates
+    
+
+    ax.plot(x_coords, y_coords, z_coords,
+             '.g', label='Observed trajectory')
 
     #####################
     # Part 3:
     # Use the initial conditions and propagation matrix for prediction
     #####################
 
-    # A = ?
-    # a = ?
-    # s = ?
-
+    # Propagation Matrix
+    A = np.matrix([[1, 0, 0,         dt,          0,          0],
+                   [0, 1, 0,          0,         dt,          0],
+                   [0, 0, 1,          0,          0,         dt],
+                   [0, 0, 0, (1 - c*dt),          0,          0],
+                   [0, 0, 0,          0, (1 - c*dt),          0],
+                   [0, 0, 0,          0,          0, (1 - c*dt)]])
+    
+    # Matrix a is the component of the motion that is changed by gravity in 
+    #    the time interval delta-t
+    a = np.transpose(np.matrix([0, 0, 0, 0, 0, g*dt]))
+    
     # Initial conditions for s0
-    # Compute the rest of sk using Eq (1)
+    s0 = np.transpose(np.matrix([0, 0, 2, 15, 3.5, 4.0]))
 
-    # ax.plot(x_coords, y_coords, z_coords,
-    #         '-k', label='Blind trajectory')
+    # Empty matrix to be filled with predicted values
+    s_pred = np.matrix(np.empty(shape=(6, (K))))
+
+    # Insert initial conditions into the first row
+    s_pred[:, 0] = s0
+    
+    # Compute the rest of sk using Eq (1), one column at a time
+    for i in range(1, K):
+        s_pred[:, i] = A * s_pred[:, i-1] + a
+
+    # Extract coordinate values from the top three rows
+    x_coords = np.array(s_pred)[0, :]  # extract x coordinates
+    y_coords = np.array(s_pred)[1, :]  # extract y coordinates
+    z_coords = np.array(s_pred)[2, :]  # extract z coordinates
+
+    ax.plot(x_coords, y_coords, z_coords,
+            '-k', label='Blind trajectory')
 
     #####################
     # Part 4:
     # Use the Kalman filter for prediction
     #####################
 
-    # B = ?
-    # C = ?
+    # Matrix B describes the scale of the errors
+    B = np.matrix([[bx,  0,  0,  0,  0,  0],
+                   [ 0, by,  0,  0,  0,  0],
+                   [ 0,  0, bz,  0,  0,  0],
+                   [ 0,  0,  0,bvx,  0,  0],
+                   [ 0,  0,  0,  0,bvx,  0],
+                   [ 0,  0,  0,  0,  0,bvz]])
+
+    # Matrix C describes the dimensional stretching
+    C = np.matrix([[rx,  0,  0, 0, 0, 0],
+                   [ 0, ry,  0, 0, 0, 0],
+                   [ 0,  0, rz, 0, 0, 0]])
```

Comment:
Same for B and C, this could save you time!
 B = np.matrix(np.zeros([6,6]))
 np.fill_diagonal(B,[bx,by,bz,bvx,bvy,bvz])

C = np.matrix(np.zeros([3,6]))
np.fill_diagonal(C,[rx,ry,rz])